File: backend/src/main/java/dev/anthonybruno/gymbuddy/auth/DbAuthenticationService.java
package dev.anthonybruno.gymbuddy.auth;

import dev.anthonybruno.gymbuddy.auth.password.BcryptPasswordHasher;
import dev.anthonybruno.gymbuddy.auth.password.PasswordHasher;
import dev.anthonybruno.gymbuddy.user.UserRepository;
import dev.anthonybruno.gymbuddy.user.model.InternalUserDetails;
import dev.anthonybruno.gymbuddy.user.model.UserDetails;

import java.util.Optional;

public class DbAuthenticationService implements AuthenticationService {

    private final UserRepository userRepository = new UserRepository();
    private final PasswordHasher passwordHasher = new BcryptPasswordHasher();

    @Override
    public Optional<UserDetails> login(String username, String password) {
        Optional<InternalUserDetails> userDetails = userRepository.getUserByUsername(username);
        if (userDetails.isPresent()) {
           String hashedPw = userDetails.get().getPassword();
           if (passwordHasher.checkPassword(hashedPw, password)) {
               return Optional.of(userDetails.get());
           }
        }
        return Optional.empty();
    }
}

File: backend/src/main/kotlin/dev/anthonybruno/gymbuddy/auth/DbAuthenticationService.kt
package dev.anthonybruno.gymbuddy.auth

import dev.anthonybruno.gymbuddy.auth.password.BcryptPasswordHasher
import dev.anthonybruno.gymbuddy.user.UserRepository
import dev.anthonybruno.gymbuddy.user.UserDetails

class DbAuthenticationService : AuthenticationService {

    private val userRepository = UserRepository()
    private val passwordHasher = BcryptPasswordHasher()

    override fun login(email: String, password: String): UserDetails? {
        val userDetails = userRepository.getUserByEmail(email) ?: return null

        val hashedPw = userDetails.password
        return if (passwordHasher.checkPassword(hashedPw, password)) {
            return userDetails
        } else null

    }
}
